toolchain/semantics/semantics_ir_formatter.cpp

```diff
@@ -246,21 +245,78 @@ class NodeNamer {
   }
 
   auto AddBlockLabel(ScopeIndex scope_idx, SemanticsNodeBlockId block_id,
-                     std::string name = "") -> void {
-    if (!block_id.is_valid()) {
+                     std::string name = "",
+                     ParseTree::Node parse_node = ParseTree::Node::Invalid)
+      -> void {
+    if (!block_id.is_valid() || labels[block_id.index].second) {
       return;
     }
 
-    ParseTree::Node parse_node = ParseTree::Node::Invalid;
-    if (auto& block = semantics_ir_.GetNodeBlock(block_id); !block.empty()) {
-      parse_node = semantics_ir_.GetNode(block.front()).parse_node();
+    if (parse_node == ParseTree::Node::Invalid) {
+      if (auto& block = semantics_ir_.GetNodeBlock(block_id); !block.empty()) {
+        parse_node = semantics_ir_.GetNode(block.front()).parse_node();
+      }
     }
 
     labels[block_id.index] = {scope_idx,
                               GetScopeInfo(scope_idx).labels.AllocateName(
                                   *this, parse_node, std::move(name))};
   }
 
+  // Finds and adds a suitable block label for the given semantics node.
+  auto AddBlockLabel(ScopeIndex scope_idx, SemanticsNodeBlockId block_id,
+                     SemanticsNode node) -> void {
+    llvm::StringRef name;
+    switch (parse_tree_.node_kind(node.parse_node())) {
+      case ParseNodeKind::IfExpressionIf:
+        switch (node.kind()) {
+          case SemanticsNodeKind::BranchIf:
+            name = "if.expr.then";
+            break;
+          case SemanticsNodeKind::Branch:
+            name = "if.expr.else";
+            break;
+          case SemanticsNodeKind::BranchWithArg:
+            name = "if.expr.cont";
```

Comment:
I know it's not 4 letters, but maybe spell this one out? Seems like we don't have too much pressure on length for these names.

```suggestion
            name = "if.expr.continue";
```

Reply:
Actually, maybe we want a different name here so it doesn't resemble the naming we use with `continue;`?

```suggestion
            name = "if.expr.resume";
```

Or `after` or `merge`?

Reply:
Nevermind, these *all* seem like they might collide, so we'll rely on the `if.expr.` context anyways... `continue` is probably better to stick with. Sorry for the noise.

Reply:
Switched to `.done` for statement forms  (currently only `if`), and `.result` for expression forms.